rmw_fastrtps_cpp/include/rmw_fastrtps_cpp/TypeSupport_impl.hpp

```diff
@@ -994,13 +997,28 @@ bool TypeSupport<MembersType>::serialize(
   assert(data);
   assert(payload);
 
-  auto ser = static_cast<eprosima::fastcdr::Cdr *>(data);
-  if (payload->max_size >= ser->getSerializedDataLength()) {
-    payload->length = static_cast<uint32_t>(ser->getSerializedDataLength());
-    payload->encapsulation = ser->endianness() ==
-      eprosima::fastcdr::Cdr::BIG_ENDIANNESS ? CDR_BE : CDR_LE;
-    memcpy(payload->data, ser->getBufferPointer(), ser->getSerializedDataLength());
-    return true;
+  auto ser_data = static_cast<SerializedData *>(data);
+  if (ser_data->is_cdr_buffer) {
```

Comment:
I am a little bit confused. Why would you call `serialize` on already binary data?

Reply:
This is the case when the call comes from rmw_publish_serialized_message.

Reply:
so right now, there is no call to serialize. The raw data will simply be wrapped in `Cdr` and directly send.
I feel like, this decision of whether it's actually necessary to serialize or not should be taken outside of the serialize function.

Reply:
When you say 'right now', do you mean on the code currently on master, or on this PR.

Whenever data is sent, the serialize function of TypeSupport class is called. In the code currently on master it just copies an externally serialized buffer into the SerializedPayload (that points) to the internal history record buffer.

The code on this PR does that on some cases (namely rmw_publish_serialized) and performs serialization directly on others (rmw_publish). On the latter case we save the allocation of a temporary buffer.